src/Microsoft.AspNetCore.Rewrite/UrlRewrite/InputParser.cs

```diff
@@ -0,0 +1,212 @@
+                if (context.Current == CloseBrace)
+                {
+                    // This is just a server variable, so we do a lookup and verify the server variable exists.
+                    parameter = context.Capture();
+                    results.Add(new PatternSegment(ServerVariables.FindServerVariable(parameter)));
+                    return;
+                }
+                else if (context.Current == Colon)
+                {
+                    parameter = context.Capture();
+                    
+                    // Only 5 strings to expect here. Case sensitive.
+                    switch (parameter)
+                    {
+                        case "ToLower":
+                            {
+                                var pattern = ParseString(context);
+                                results.Add(new PatternSegment((ctx, ruleMatch, condMatch) =>
+                                {
+                                    var str = pattern.Evaluate(ctx, ruleMatch, condMatch);
+                                    return str.ToLowerInvariant();
+                                }));
+
+                                // at this point, we expect our context to be on the ending closing brace,
+                                // because the ParseString() call will increment the context until it 
+                                // has processed the new string.
+                                if (context.Current != CloseBrace)
+                                {
+                                    throw new FormatException();
+                                }
+                            }
+                            break;
+                        case "UrlDecode":
+                            {
+                                throw new NotImplementedException("Do no support UrlDecoding.");
+                            }
+                        case "UrlEncode":
+                            {
+                                var pattern = ParseString(context);
+                                results.Add(new PatternSegment((ctx, ruleMatch, condMatch) =>
+                                {
+                                    var str = pattern.Evaluate(ctx, ruleMatch, condMatch);
+                                    return UrlEncoder.Default.Encode(str);
+                                }));
+
+                                if (context.Current != CloseBrace)
+                                {
+                                    throw new FormatException();
+                                }
+                            }
+                            break;
+                        case "R":
+                            {
+                                var index = GetBackReferenceIndex(context);
+                                results.Add(new PatternSegment((ctx, ruleMatch, condMatch) =>
+                                {
+                                    return ruleMatch?.Groups[index]?.Value;
```

Comment:
The index will never be null, and we always want to add it to the results. As the func is applied on obtaining the request, we don't know that the backreference exists until the regex is applied.